### DISCUSSION TITLE
Final check before publishing PG 1.0.0

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#62] [General]
Ensure that future changes do not break compatibility (currently valid documents must never become invalid due to extensions).

**Comment by hchiba1** [comment]
**Comments:**
```
a key:2#comment
```
is valid now, and it is same as
```
a key:2 #comment
```
i.e., comments without spaces are allowed in some edge cases.

Is the former example permanently valid?
Make sure that future changes will not invalidate the former example.

**Reply by hchiba1** [reply]
```
a key:#comment
  val
```
is also valid now.

This should not be invalid in the future.

**Reply by nichtich** [reply]
I've added a test case to make sure it stays valid (although it's not beautiful): https://github.com/pg-format/pg-test-suite/commit/ef4733e9cef456adad4886e495926be6b660977d

**Comment by hchiba1** [comment]
Is it possible to support condenced forms of edges in the future, without breaking the current valid examples?
```
a--b
c->d
```
Or will we give up condensed forms permanently?

**Reply by nichtich** [reply]
As `-` is no special character, `a--b` as edge cannot be distinguished from node with id `a--b`. Directed edges, however could be condensed as `c->d` has no valid alternative.

**Reply by hchiba1** [reply]
That would, in effect, mean giving up on condensed forms of edges (instead allowing `--` in URIs).

**Comment by hchiba1** [comment]
**Disallowing space between the property key and the colon**
```
a k :v
```

**Reply by hchiba1** [reply]
This rule was set mainly for syntax highlighting.

**Reply by hchiba1** [reply]
Is it ever possible to admit that in the future?
(To avoid confusion, it would be better to disallow it even in the future.)

**Reply by nichtich** [reply]
No, this seems to be not detected as invalid by syntax highlighting but the grammar forbids `a k :v`.

**Comment by hchiba1** [comment]
Commas are allowed after the second character of unquoted IDs and property keys, permanently.
```
a,a k,k:v,w
```

**Reply by nichtich** [reply]
This is parsed by the implementation (included in the test suite) but not by syntax highlighter. Ticket: https://github.com/pg-format/pg-highlight/issues/6

**Comment by hchiba1** [comment]
Colons are allowed after the second character of unquoted IDs and property values, permanently.
```
a:a k:v
```
```
a k:v:
```

**Reply by hchiba1** [reply]
```
node a:b:c
```
is same as 
```
node a: b:c
```
and different from
```
node a:b: c
```

**Comment by hchiba1** [comment]
`#` is allowed after the second character of unquoted IDs, labels, property keys and values, permanently.

**Comment by hchiba1** [comment]
There are four places for unquoted strings:
- node identifiers
- labels
- property keys
- property values

`#x00-#x20` `<` `>` `"` `{` `}` `|` `^` ``\`` are disallowed in unquoted strings.
There remains a possibility to reduce the disallowed characters.

`'` `"` `:` `#` `,` `-` are disallowed at the first position of the unquoted strings.
There remains a slight possibility to reduce the disallowed characters.

**Reply by hchiba1** [reply]
`;` is now allowed in unquoted strings, which means that `;` will not be used as an additional line separator in future extension.

**Reply by hchiba1** [reply]
`'` is allowed after the second character of unquoted IDs, labels, property keys and values, permanently.

**Comment by hchiba1** [comment]
**Escape**
`/` is included in `Escaped` rule, thus
```
a k:"\/"
```
is valid.
In RDF 1.1 Turtle, `ECHAR` does not include `/`,
If `/` is ever removed from `Escaped` rule, this example PG become invalid.

**Reply by hchiba1** [reply]
`\v` is not included in `Escaped`, thus
```
a k:"\v"
```
is invalid.

**Reply by nichtich** [reply]
Both SPARQL and Turtle have these escape characters: `[tbnrf"'\]`

PG Format has these escape characters: `[tbnrf"'\/]`

Any other character after `\` is invalid but adding more characters such as `v` in future versions would be backwards compatible.

**Reply by hchiba1** [reply]
Was the inclusion of`\/` in `Escaped` intended?

**Comment by hchiba1** [comment]
Quoted strings can include newlines (`#x0A`).

**Comment by hchiba1** [comment]
**Code point**
Code point after `U+10000` (such as `U+1D11E`) should be represented by surrogate pairs (such as `\uD834\uDD1E`).

**Reply by hchiba1** [reply]
In that sense, the `Codepoint` rule in the grammar does not strictly represent all code points.
(Maybe it's just a matter of the variable name.)

**Reply by nichtich** [reply]
Handling of code points above `U+10000` is inconvenient indeed, but that's how JSON does it. I think adding escape sequence `\U` like [in Turtle](https://www.w3.org/TR/turtle/#sec-escapes) makes sense. But this could also be [postponed to version 1.1.0](https://github.com/pg-format/specification/issues/15)?

I've [changed the grammar to not use `Codepoint`](https://github.com/pg-format/specification/commit/0da46a9e451650462ceee85b24145abf17ac83d1#diff-142d492309aec76ce25caedd8d7eb0bdfe11ecdfa91f60eb015f56a406a41446L441).

**Comment by hchiba1** [comment]
PG with no statements
```
# empty
```
is valid.

**Comment by hchiba1** [comment]
**Number** can be in exponential form.
```
a k:1.0e+2
```
is same as 
```
a k:100
```

**Comment by hchiba1** [comment]
Empty string node identifiers are disallowed.
```
"" k:v
```
It's easy to allow this grammatically. However, it isn't very meaningful, so we disallow it.

**Comment by hchiba1** [comment]
Can `ValueList` be an empty list? 
=> No, `ValueList` should not be an empty list. 
    (Although it may be easy to support an empty list of values in PG-JSON(L), it might be difficult and confusing to support in PG.)

**Reply by nichtich** [reply]
The PG format model can neither model empty lists nor one-element-lists (there is no difference in one property value and a list with one element as property value). This is a design choice: property values are always non-empty lists. Empty list could be allowed with syntax extension: https://github.com/pg-format/specification/issues/16
